src/Core/functions.php

```diff
@@ -185,11 +185,11 @@ function pj(mixed $var): mixed
      * environment information.
      *
      * @param string $key Environment variable name.
-     * @param string|bool|null $default Specify a default value in case the environment variable is not defined.
-     * @return string|float|int|bool|null Environment variable setting.
+     * @param array|string|float|int|bool|null $default Specify a default value in case the environment variable is not defined.
+     * @return array|string|float|int|bool|null Environment variable setting.
      * @link https://book.cakephp.org/4/en/core-libraries/global-constants-and-functions.html#env
      */
-    function env(string $key, string|float|int|bool|null $default = null): string|float|int|bool|null
+    function env(string $key, array|string|float|int|bool|null $default = null): array|string|float|int|bool|null
```

Comment:
I am wondering if the return type should match that of PHP's `getenv()` which is `string|array|false`.

Reply:
Did we have this typed expecting some auto-conversion? I don't think that's possible now with strict types.

Reply:
I don't think auto type conversion was expected. Int or float return types would probably be possible only for user set/overriding values.

Reply:
hmm https://www.php.net/manual/en/function.putenv.php requires a string. 🤷

Reply:
Haha, it's just one argument, in 'a=b' form.

Reply:
I think it's because we pull from $_ENV first. Users could set a value in $_ENV directly with a type.
```php
        $val = $_SERVER[$key] ?? $_ENV[$key] ?? null;
        assert($val === null || is_scalar($val));
        if ($val == null && getenv($key) !== false) {
            $val = (string)getenv($key);
        }
```

Reply:
> I am wondering if the return type should match that of PHP's getenv() which is string|array|false.

Our `env` method reads from more than just the environment vars though. It also reads from `_SERVER` which can contain float values (request_time_start). If we narrow the types, we'll get fatal errors for these keys :cry:

Reply:
NVM then.

Reply:
Feels like we should just return mixed and not bother with it.

Reply:
The typing failures in this build, are indicative of the type errors that will spill out into userland code should we make this change. There are very few array values in `$_SERVER` and none in `$_ENV`. If `env()` should work more like `getenv()` then perhaps we make the known array values into runtime errors and don't support the scenario of `env('argv')` which is the only array value in my local `php`'s `$_SERVER` value 🤷

Reply:
> we make the known array values into runtime errors and don't support the scenario of env('argv') which is the only array value in my local php's $_SERVER value 🤷

:+1: